### PR TITLE
[README] Add the readme for the run.sh in aishsll asr1

### PR DESCRIPTION
**Pull request by Jackwaterveg**
### PR types
Others 

### PR changes
Docs

### Describe
WIP样板间： 给aishell的asr1编写README

### CODE REVIEW

File: examples/aishell/asr1/READEME.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
+# Conformer ASR with Aishell
+
+This example contains code used to train a [Conformer](http://arxiv.org/abs/2008.03802) model with [Aishell dataset](http://www.openslr.org/resources/33)
+
+## Overview
+
+All the scirpt you need is in the ```run.sh```. The are several stages in the ```run.sh```. You can choose to run some stages by set the ```stage``` and ```stop_stage ``` . For xample , if you run this script:
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: There are several stages

Comment (yt605155624) [reply examples/aishell/asr1/READEME.md]: For example

File: examples/aishell/asr1/READEME.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+```
+bash run.sh --stage 2 --stop_stage 3
+```
+
+the script will execute the code in stage 2 and stage 3.
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: the -> This

File: examples/aishell/asr1/READEME.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+This doc will decribe the scripts in the ```run.sh``` in detail.
+
+
+
+## The environment variables
+
+The path.sh contains the environment variable. 
+```
+sourth path.sh
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: source

File: examples/aishell/asr1/READEME.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+If you only want to process the data. You can run
+
+```
+bash run.sh --stage 0 --stop_stage 0
+```
+
+or you can also just run these scripts in your command line.
+
+```
+sourth path.sh
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: source

File: examples/aishell/asr1/READEME.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+For example:
+
+```
+bash run.sh --gpus 0,1 --avg_num 20
+```
+
+This script will set the ```gpus``` and ``avg_num`` when you use the command line.
+
+## Stage 0: process data
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: 不同 stage 大小写保持一致 process -> Process

File: examples/aishell/asr1/READEME.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+The ```avg.sh``` is in the ```../../../utils/``` which is define in the ```path.sh```.
+If you want to get the final model,  you can use the script below to execute the stage 0, stage 1, and stage 2:
+
+```
+bash run.sh --stage 0 --stop_stage 2
+```
+
+or you can run these scripts in command line (only use CPU).
+```
+sourth path.sh
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: source

File: examples/aishell/asr1/READEME.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+```
+bash run.sh --stage 0 --stop_stage 2
+bash run.sh --stage 4 --stop_stage 4
+```
+
+or you can also use these scripts in command line (only use CPU).
+
+```
+sourth path.sh
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: source

File: examples/aishell/asr1/READEME.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+You need to prepare an audio file, please confirme the sample rate of the audio is 16K.  Assume the path of the audio file is ```data/test_audio.wav```,  you can get the result by runing the script below.
+
+```
+CUDA_VISIBLE_DEVICES= ./local/test_hub.sh conf/conformer.yaml exp/conformer/checkpoints/avg_20 data/test_audio.wav
+```
+
+
+
+## Stage 51： Export the static model （not supported at now）
```
Comment (yt605155624) [inline examples/aishell/asr1/READEME.md]: 51? 换个说法是不是更好一点呀

File: examples/aishell/asr1/READEME.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+CUDA_VISIBLE_DEVICES= ./local/train.sh conf/conformer.yaml conformer
+avg.sh best exp/conformer/checkpoints 20
+CUDA_VISIBLE_DEVICES= ./local/test.sh conf/conformer.yaml exp/conformer/checkpoints/avg_20
+```
+
+
+
+## Stage 4: Get ctc alignment of test data
+
+If you want to get the alignment between the audio and the text, you can you the ctc alignment. The code of this stage is shown below:
```
Comment (xiegegege) [inline examples/aishell/asr1/READEME.md]: you can use?

File: examples/aishell/asr1/READEME.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+ fi
+```
+
+If you want to train a model, test it and do the alignment,  you can use the script below to execute the stage 0, stage 1,  stage 2, and stage 3 :
+
+```
+bash run.sh --stage 0 --stop_stage 4
+```
+
+or if you only need to train a model and do the aliggnment, you can use these scripts to escape the stage 3(test stage):
```
Comment (xiegegege) [inline examples/aishell/asr1/READEME.md]: alignment拼错了

File: examples/aishell/asr1/run.sh [hunk 11]
```diff
@@ -41,18 +41,18 @@ if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
...
     CUDA_VISIBLE_DEVICES=0 ./local/test_hub.sh ${conf_path} exp/${ckpt}/checkpoints/${avg_ckpt} ${audio_file} || exit -1
 fi
 
-if [ ${stage} -le 7 ] && [ ${stop_stage} -ge 7 ]; then
+# if [ ${stage} -le 51 ] && [ ${stop_stage} -ge 51 ]; then
+#     # export ckpt avg_n
+#     CUDA_VISIBLE_DEVICES=0 ./local/export.sh ${conf_path} exp/${ckpt}/checkpoints/${avg_ckpt} exp/${ckpt}/checkpoints/${avg_ckpt}.jit
+# fi
+
+if [ ${stage} -le 101 ] && [ ${stop_stage} -ge 101 ]; then
```
Comment (yt605155624) [inline examples/aishell/asr1/run.sh:56]: 如果默认的 stop_stage 是 50 ，那这里改成 52？

File: examples/aishell/asr1/READEME.md [hunk 12]
```diff
@@ -0,0 +1,301 @@
+# Conformer ASR with Aishell
+
+This example contains code used to train a [Conformer](http://arxiv.org/abs/2008.03802) model with [Aishell dataset](http://www.openslr.org/resources/33)
+
+## Overview
+
+All the scirpt you need is in the ```run.sh```. There are several stages in the ```run.sh```, and each stage has it's function.
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: scirpt -> scripts
it's -> its

File: examples/aishell/asr1/READEME.md [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+
+| Stage | Function                                                     |
+| :---- | :----------------------------------------------------------- |
+| 0     | Process data. It includes: <br>       (1) Download the dataset <br>       (2) Caculate the CMVN of the train dataset <br>       (3) Get the vocabulary file <br>       (4) Get the manifest files of the train, development and test dataset |
+| 1     | Train the model                                              |
+| 2     | Get the final model by average the top-k model , set k = 1 means choose the best model |
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: by average the top-k model -> by averaging the top-k models

File: examples/aishell/asr1/READEME.md [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+| 3     | Test the final model performance                             |
+| 4     | Get ctc alignment of test data using the final model         |
+| 5     | Infer the single audio file                                  |
+| 51    | (Not supported at now) Transform the dynamic graph model to static graph model |
+| 101   | (Need further installation) Train language model and Build TLG |
+
+You can choose to run a range of  stages by set the ```stage``` and ```stop_stage ``` .
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: a range of  stages by set the ```stage``` and ```stop_stage ``` . -> a range of  stages by setting ```stage``` and ```stop_stage ```.

File: examples/aishell/asr1/READEME.md [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+| 2     | Get the final model by average the top-k model , set k = 1 means choose the best model |
+| 3     | Test the final model performance                             |
+| 4     | Get ctc alignment of test data using the final model         |
+| 5     | Infer the single audio file                                  |
+| 51    | (Not supported at now) Transform the dynamic graph model to static graph model |
+| 101   | (Need further installation) Train language model and Build TLG |
+
+You can choose to run a range of  stages by set the ```stage``` and ```stop_stage ``` . 
+
+For example , if you want to execute the code in stage 2 and stage 3, you can run this script:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: For example , -> For example,

File: examples/aishell/asr1/READEME.md [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+Or you can set ```stage``` equal to ```stop-stage``` to only run one stage.
+For example, if you only want to run ```stage 0```, you can use the script below:
+
+```bash
+bash run.sh --stage 0 --stop_stage 0
+```
+
+
+
+The document below will decribe the scripts in the ```run.sh``` in detail.
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: decribe -> describe

File: examples/aishell/asr1/READEME.md [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+
+```bash
+bash run.sh --gpus 0,1 --avg_num 20
+```
+
+
+
+## Stage 0: Data processing
+
+To use this example, you need to process data firstly and  you can use the stage 0 in the ```run.sh``` to do this. The code is shown below:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: and  you can use the stage 0 -> and you can use stage 0

File: examples/aishell/asr1/READEME.md [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+
+You can also just run these scripts in your command line.
+
+```bash
+source path.sh
+source ${MAIN_ROOT}/utils/parse_options.sh
+bash ./local/data.sh
+```
+
+Aftre processing the data, the ``data`` directory will be look like this:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: Aftre - > After
be look like -> look like

File: examples/aishell/asr1/READEME.md [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+source ${MAIN_ROOT}/utils/parse_options.sh
+bash ./local/data.sh
+CUDA_VISIBLE_DEVICES= ./local/train.sh conf/conformer.yaml  conformer
+```
+
+
+
+## Stage 2:  Top-k model averaging
+
+After training the model,  we need to get the final model for test and infer. In every epoch, the model checkpoint is saved , so we can choose the best model from them based on the validation loss or we can sort them and average the top-k model parameters to get the final model.  We can use the stage 2 to do this, and the code is shown below:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: for test and infer -> for testing and inference
is saved , -> is saved,
the stage 2 -> stage 2 一般具体的名字前面可以不加定冠词，除非换成序数词表达如The 2nd stage。

File: examples/aishell/asr1/READEME.md [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+
+To use this example, you need to process data firstly and  you can use the stage 0 in the ```run.sh``` to do this. The code is shown below:
+```bash
+ if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+     # prepare data
+     bash ./local/data.sh || exit -1
+ fi
+```
+
+The stage 0 is for processing the data.
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: The stage 0 -> Stage 0

File: examples/aishell/asr1/READEME.md [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+bash ./local/data.sh
+CUDA_VISIBLE_DEVICES= ./local/train.sh conf/conformer.yaml conformer
+avg.sh best exp/conformer/checkpoints 20
+```
+
+
+
+## Stage 3: Model Testing
+
+To know the preformence of the model, test stage is needed. The code of test stage is shown below:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: 被动语态不建议用得太多哈
To know the preformence of the model, test stage is needed. -> 
The test stage is to evaluate the model performance.

File: examples/aishell/asr1/READEME.md [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+If you want to get the alignment between the audio and the text, you can use the ctc alignment. The code of this stage is shown below:
+
+```bash
+ if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
+     # ctc alignment of test data
+     CUDA_VISIBLE_DEVICES=0 ./local/align.sh ${conf_path} exp/${ckpt}/checkpoints/${avg_ckpt} || exit -1
+ fi
+```
+
+If you want to train the model, test it and do the alignment,  you can use the script below to execute the stage 0, stage 1,  stage 2, and stage 3 :
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: the stage 0 -> stage 0
还有注意逗号紧跟在单词后，和下一个单词之间仅留一个空格。

File: examples/aishell/asr1/READEME.md [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+```
+
+or if you only need to train a model and do the alignment, you can use these scripts to escape the stage 3(test stage):
+
+```bash
+bash run.sh --stage 0 --stop_stage 2
+bash run.sh --stage 4 --stop_stage 4
+```
+
+or you can also use these scripts in command line (only use CPU).
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: in command line -> in the command line

File: examples/aishell/asr1/READEME.md [hunk 24]
```diff
@@ -0,0 +1,301 @@
...
+# test stage is optional
+CUDA_VISIBLE_DEVICES= ./local/test.sh conf/conformer.yaml exp/conformer/checkpoints/avg_20
+CUDA_VISIBLE_DEVICES= ./local/align.sh conf/conformer.yaml exp/conformer/checkpoints/avg_20
+```
+
+
+
+## Stage 5: Single audio file inference
+
+In some situation, you want to use the trained model to do the inference for the single audio file. You can use the stage  5. The code is shown below
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: situation -> situations
the stage  5 -> stage 5

File: examples/aishell/asr1/READEME.md [hunk 25]
```diff
@@ -0,0 +1,301 @@
...
+
+```bash
+ if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ]; then
+     # test a single .wav file
+     CUDA_VISIBLE_DEVICES=0 ./local/test_hub.sh ${conf_path} exp/${ckpt}/checkpoints/${avg_ckpt} ${audio_file} || exit -1
+ fi
+```
+
+you can train the model by yourself using ```bash run.sh --stage 0 --stop_stage 3```， or you can download the pretrained model by the script below:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: ， or you -> , or you 
这里逗号是中文逗号

File: examples/aishell/asr1/READEME.md [hunk 26]
```diff
@@ -0,0 +1,301 @@
...
+
+```
+wget https://deepspeech.bj.bcebos.com/release2.1/aishell/s1/aishell.release.tar.gz
+tar aishell.release.tar.gz
+```
+
+You need to prepare an audio file, please confirme the sample rate of the audio is 16K.  Assume the path of the audio file is ```data/test_audio.wav```,  you can get the result by runing the script below.
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: confirme -> confirm
16K.  Assume -> 16K. Assume 句号后多了一个空格
,  you can get the result by runing the -> , you can get the result by running the
by runing the script below -> by running the script below

File: examples/aishell/asr1/READEME.md [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+ fi
+```
+
+It is not supported at now, so we set a large stage number for this stage.
+
+
+
+## Stage: 101 Language model training and TLG building (Need further installation! )
+
+You need to install the kaldi and srilm to use the stage 101,  it is used for train language model and build TLG. To do further installation, you need to do these:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: use the stage 101,  it is used for train language model and build TLG. -> use stage 101, it is used for training language model and building TLG.

File: examples/aishell/asr1/READEME.md [hunk 28]
```diff
@@ -0,0 +1,346 @@
...
+| :---- | :----------------------------------------------------------- |
+| 0     | Process data. It includes: <br>       (1) Download the dataset <br>       (2) Caculate the CMVN of the train dataset <br>       (3) Get the vocabulary file <br>       (4) Get the manifest files of the train, development and test dataset |
+| 1     | Train the model                                              |
+| 2     | Get the final model by averaging the top-k models, set k = 1 means choose the best model |
+| 3     | Test the final model performance                             |
+| 4     | Get ctc alignment of test data using the final model         |
+| 5     | Infer the single audio file                                  |
+
+
+You can choose to run a range of  stages by setting the ```stage``` and ```stop_stage ``` .
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: of和stages间的空格删掉一个，``stage``和``stop_stage``前面的the建议删掉。

File: examples/aishell/asr1/READEME.md [hunk 29]
```diff
@@ -0,0 +1,346 @@
...
+For example, if you only want to run ```stage 0```, you can use the script below:
+
+```bash
+bash run.sh --stage 0 --stop_stage 0
+```
+
+
+
+The document below will describe the scripts in the ```run.sh``` in detail.
+
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md:37]: the ```run.sh``的the可以去掉

File: examples/aishell/asr1/READEME.md [hunk 30]
```diff
@@ -0,0 +1,346 @@
...
+## The Environment Variables
+
+The path.sh contains the environment variable.
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: variable - > variables

File: examples/aishell/asr1/READEME.md [hunk 31]
```diff
@@ -0,0 +1,346 @@
...
+The path.sh contains the environment variable. 
+
+```bash
+source path.sh
+```
+
+This script needs to be run firstly.  
+
+And another script is also needed:
+
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md:47]: 这两句可以合并哈

File: examples/aishell/asr1/READEME.md [hunk 32]
```diff
@@ -0,0 +1,346 @@
...
+bash ./local/data.sh
+CUDA_VISIBLE_DEVICES= ./local/train.sh conf/conformer.yaml conformer
+avg.sh best exp/conformer/checkpoints 20
+```
+
+
+
+## Stage 3: Model Testing
+
+The test stage is to evaluate the model performance.. The code of test stage is shown below:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: .. -> .
test stage -> the test stage

File: examples/aishell/asr1/READEME.md [hunk 33]
```diff
@@ -0,0 +1,346 @@
...
+
+```bash
+ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
+     # test ckpt avg_n
+     CUDA_VISIBLE_DEVICES=0 ./local/test.sh ${conf_path} exp/${ckpt}/checkpoints/${avg_ckpt} || exit -1
+ fi
+```
+
+If you want to train a model and test it,  you can use the script below to execute stage 0, stage 1,  stage 2, and stage 3 :
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: 一般我们只用一个空格哈

File: examples/aishell/asr1/READEME.md [hunk 34]
```diff
@@ -0,0 +1,346 @@
...
+In some situations, you want to use the trained model to do the inference for the single audio file. You can use stage  5. The code is shown below
+
+```bash
+ if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ]; then
+     # test a single .wav file
+     CUDA_VISIBLE_DEVICES=0 ./local/test_wav.sh ${conf_path} exp/${ckpt}/checkpoints/${avg_ckpt} ${audio_file} || exit -1
+ fi
+```
+
+you can train the model by yourself using ```bash run.sh --stage 0 --stop_stage 3```, or you can download the pretrained model by the script below:
```
Comment (Mingxue-Xu) [inline examples/aishell/asr1/READEME.md]: by -> through

**Review by yt605155624 — APPROVED** [review]
LGTM

**Pull request merged**